source-code/core2/src/plugin/api.ts

```diff
@@ -0,0 +1,86 @@
+import type { InlangApp } from "../app/api.js"
+import type { LanguageTag } from "../languageTag.js"
+import type { LintRule } from "../lint/api.js"
+import type { Message } from "../messages/schema.js"
+
+/**
+ * Proposal 1
+ *
+ * Setup function everything in one.
+ *
+ * + simple to use
+ * - no access to static props.
+ *   if a plugin crashes during setup -> complicated to show plugin with "id" crashed.
+ */
+export type Plugin_Proposal_1<Options extends Record<string, string> = Record<string, never>> =
+	(args: { options: Options; inlang: InlangApp }) => {
+		id: `${string}.${string}`
+		displayName: string
+		extendLanguageTags?: () => LanguageTag[]
+		loadMessages?: () => Message[]
+		saveMessages?: (args: { messages: Message[] }) => void
+		addLintRules?: () => LintRule[]
+	}
+
+export const example1: Plugin_Proposal_1<{ pathPattern: string }> = ({ options, inlang }) => {
+	return {
+		id: "inlang.myPlugin",
+		displayName: "My Plugin",
+		extendLanguageTags: () => {
+			return ["en-US"]
+		},
+		loadMessages: () => {
+			for (const languageTag of inlang.config.languageTags) {
+				console.log(languageTag + options.pathPattern)
+			}
+			return []
+		},
+		saveMessages: ({ messages }) => {
+			console.log(messages)
+		},
+	}
+}
+
+/**
+ * Proposal 2
+ *
+ * Split static meta data from setup function.
+ *
+ * + access to static without crash chance
+ * - arguably not so nice API.
+ */
+export type Plugin_Proposal_2<Options extends Record<string, string> = Record<string, never>> = {
+	meta: {
+		id: `${string}.${string}`
+		displayName: string
+	}
+	setup: (args: { options: Options; inlang: InlangApp }) => {
+		extendLanguageTags?: () => LanguageTag[]
+		loadMessages?: () => Message[]
+		saveMessages?: (args: { messages: Message[] }) => void
+		addLintRules?: () => LintRule[]
```

Comment:
Wouldn't it be changeLintRules?
Can be plugins be treated the same way? No, because they would replace themselves. 
Do you guys see a solution for that? Maybe out of scope for now. But I see a settings menu in the editor where I could literaly configure the config. Which would be nice 👍

Reply:
The API is runtime context. 

The API DOES NOT extend the config. Hence, you can't remove lint rules. You only add lint rules.

Reply:
What is this rule then doing?

Reply:
It adds lint rules to the inlang instance which can be processed by the instance linter.